### ISSUE TITLE
Can we speed up the getMetricIDsForDateTagFilter function by passing metricIDs?

### ISSUE THREAD

**Issue by faceair** [VictoriaMetrics/VictoriaMetrics#781]
I'm learning the code for VictoriaMetrics.

In the `getMetricIDsForDateAndFilters` function, the loop calls `getMetricIDsForDateTagFilter` to get the match result for each `tagFilter.`

![image](https://user-images.githubusercontent.com/3659110/93714549-02de2300-fb96-11ea-8ab8-77e9c512be8e.png)

But why not pass the metricIDs directly to `getMetricIDsForDateTagFilter`?   This should quickly bypass some of the heavy regular expression matching.

![image](https://user-images.githubusercontent.com/3659110/93714756-543ae200-fb97-11ea-957d-3af616df7574.png)

**Comment by valyala** [comment]
This approach looks promising! @faceair , could you prepare a pull request for further investigation? Make sure it properly handles both positive and negative matches. I.e. `foo{bar!="baz"}` and `foo{bar!~"baz.*"}` shouldn't break after the change.

**Comment by faceair** [comment]
Yes, I know the difference between the two matches and I try to implement it.

**Comment by faceair** [comment]
I implemented the first version, but I now have another idea.
After we get the metricIDs with getMetricIDsForDateTagFilter for the first time, there is no need to keep using getMetricIDsForDateTagFilter to do the rest of the matching.
We can use the index of MetricIDToMetricName to find all the metrics and then cycle through the collection for the rest of the TagFilter matches, which should be more efficient.

**Comment by valyala** [comment]
> We can use the index of MetricIDToMetricName to find all the metrics and then cycle through the collection for the rest of the TagFilter matches, which should be more efficient.

Unfortunately this doesn't work, since matching by metricName requires the following steps:
1) locating and unmarshaling all the metricName values from the inverted index for the given metricIDs;
2) matching found metricNames against all the tag filters.

These steps are usually much slower than locating and intersecting metricIDs by tag filters.

**Comment by faceair** [comment]
Yes, I know the two steps you're talking about. But now isn't getMetricIDsForDateTagFilter lookup also divided into these two steps ?
Is it because a single index lookup has a higher overhead ?
The way I'm talking about would do more index lookups and perform fewer tag matches.

**Comment by valyala** [comment]
> But now isn't getMetricIDsForDateTagFilter lookup also divided into these two steps ?

Yes, but metricNames lookup is performed only when the number of metricIDs is relatively small.

> Is it because a single index lookup has a higher overhead ?

Yes. Performance difference between sequential index scan and index lookup is multiple orders of magnitude (i.e. 1000x and more). So it is better to scan 100M of metricIDs than to look up 10K metricNames.

**Comment by faceair** [comment]
Thank you for the detailed explanation, I understand. Looks like it's the best way to go for now.

**Comment by valyala** [comment]
FYI, enhancements mentioned above have been included in [v1.41.1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) . This should improve performance for queries with complex regexp filters.

**Comment by faceair** [comment]
There should be performance improvements not just for complex regular matching. I think this PR is more important for pruning.

I see that the code is now sorting the TagFilter. When we run the smallest TagFilter, there are a lot of metrcisIDs left behind, and when we run the next TagFilter, we only need to match the metrcisIDs from the previous round of matches. As we iterate down, there are fewer and fewer metrcisIDs to match.

Also we should be able to continue to optimize the sorting rules here so that this is sorted by true match overhead.
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/v1.43.0/lib/storage/index_db.go#L2584-L2590

When a regular match occurs, we score the regular expression to see how many times its overhead is the normal match. Then we multiply this overhead by the number of metrics to get the true overhead of this TagFilter.

For example, when we have 1000 metrics that contain the {a=bcc} label.

```
filter1 := TagFilter{
 Label: "a"
 Value: "bcc"
 Count: 101
}
filter2 :=TagFilter{
 Label: "a"
 Re: "^b.+?"
 Count: 100
}
```

In the current implementation, we would execute filter2 first, but the real case is that regular matches will be executed 100 times when this regular expression is complex, which may be more overhead than filter1 executing 101 times which equals judgment. But if we introduce a scoring mechanism and we evaluate the overhead of a true regular match, it would be.

filter1 score = 101
filter2 score = 100 * 100 (regex complexity)

We can perform low-cost filter pruning of metricsIDs.

**Comment by valyala** [comment]
@faceair , the idea with weighted sorting for tag filters looks promising. Could you try implementing it and providing the corresponding pull request?

**Comment by faceair** [comment]
Okay, I'll try.

**Comment by valyala** [comment]
Thanks! Merged the pull request.

**Comment by valyala** [comment]
FYI, the enhancement that sorts label filters according to their weights has been included in [v1.45.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
